modules/c++/mem/include/mem/ScratchMemory.h

```diff
@@ -0,0 +1,140 @@
+/* =========================================================================
+ * This file is part of mem-c++
+ * =========================================================================
+ *
+ * (C) Copyright 2004 - 2018, MDA Information Systems LLC
+ *
+ * mem-c++ is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef __MEM_SCRATCHMEMORY_H__
+#define __MEM_SCRATCHMEMORY_H__
+
+#include <stddef.h>
+#include <map>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <except/Exception.h>
+#include <mem/BufferView.h>
+#include <sys/Conf.h>
+
+namespace mem
+{
+
+/*!
+ *  \class ScratchMemory
+ *  \brief Handle reservation of scratch memory segments within a single buffer.
+ *
+ *  Memory segments should be reserved during a "setup" phase using the put
+ *  method. Once all segments are reserved, call the setup method to set up
+ *  the underlying memory and ensure the alignment requirements of each segment.
+ *  The get method may be used afterwards to obtain pointers to the memory
+ *  segments.
+ */
+class ScratchMemory
+{
+public:
+    ScratchMemory() :
+        mNumBytesNeeded(0),
+        mIsSetup(false)
+    {
+    }
+
+    /*!
+     * \brief Reserve a buffer segment within this scratch memory buffer.
+     *
+     * \param key Identifier for scratch segment
+     * \param numBytes Size of scratch buffer
+     * \param numBuffers Number of distinct buffers to set up
+     * \param alignment Number of bytes to align segment pointer
+     *
+     * \throws except::Exception if the given key has already been used
+     */
+    template <typename T>
+    void put(const std::string& key,
+             size_t numBytes,
```

Comment:
Rename this to numElements since you are going to multiply it by sizeof(T) later on.

Reply:
Done.